### PR TITLE
Fix the documentation for Base.(===)

### LINKED ISSUE #49254: Base.(===) doesn't appear to obey its specification
I don't understand `===`.  [The doc](https://docs.julialang.org/en/v1/base/base/#All-Objects) says "First the types of `x` and `y` are compared. If those are identical, mutable objects are compared by address in memory and immutable objects (such as numbers) are compared by contents at the bit level."  But...

```
julia> t1 = ("a", 1)
("a", 1)

julia> t2 = ("a", 1)
("a", 1)

julia> pointer(t1[1]) == pointer(t2[1])
false

julia> t1 === t2
true

julia> versioninfo()
Julia Version 1.8.2
Commit 36034abf260 (2022-09-29 15:21 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.3.0)
  CPU: 12 × Apple M2 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_SSL_CA_ROOTS_PATH = 
```
It appears `t1 === t2` should be `false`, as they are not bitwise equal (they contain different pointers).

Either the documentation is wrong (a bug that should be fixed) or the implementation is in error.  I hope it is the documentation, which I suspect should say "First the types of `x` and `y` are compared. If those are identical, mutable objects are compared by address in memory, primitive objects (such as numbers) are compared by contents at the bit level, and immutable objects are compared field-by-field using `===`."

### PR DESCRIPTION
**Pull request by gafter**
Fixes #49254

### CODE REVIEW

**Review by quinnj — APPROVED** [review]

File: base/operators.jl [hunk 1]
```diff
@@ -282,9 +282,10 @@ const ≠ = !=
 
 Determine whether `x` and `y` are identical, in the sense that no program could distinguish
 them. First the types of `x` and `y` are compared. If those are identical, mutable objects
-are compared by address in memory and immutable objects (such as numbers) are compared by
-contents at the bit level. This function is sometimes called "egal".
-It always returns a `Bool` value.
+are compared by address in memory, primitive objects (such as numbers) are compared by
```
Comment (oscardssmith) [inline base/operators.jl:285]: I know this is somewhat pre-existing, but giving "numbers" as an example is somewhat misleading (e.g. `Complex` or `BigFloat`).

**Comment by Seelengrab** [comment]
I'm not sure this is more correct, is it? The original issue had Strings, which are considered egal if their bitwise contents are the same since they are (technically) immutable:

```
julia> a = "a"
"a"

julia> b = "a"
"a"

julia> a === b
true

julia> pointer(a)
Ptr{UInt8} @0x00007fea71731200

julia> pointer(b)
Ptr{UInt8} @0x00007fea71731650

julia> pointer(a) == pointer(b)
false
```

Whether or not the pointers to them (which sholdn't really be possible to get, as they are technically immutable, even though it's necessary for technical reasons) are equal or not shouldn't matter for them.

File: base/operators.jl (hunk 1 above)
Comment (gafter) [reply base/operators.jl]: I think the latest revision addresses this.

**Comment by gafter** [comment]
I think I've addressed all comments.

**Comment by Seelengrab** [comment]
What I mean is that as far as egality is concerned, Strings are no different from other immutables - only their bitwise contents matter. That only a pointer is stored in the tuple is an implementation detail.

Saying that `===` is specialized for some types is a bit misleading, since you can't specialize `===` for new types due to being built-in and not generic. Further, comparing non-mutables field-by-field also isn't quite correct - new `primitive type`s don't have any fields (and you can't add any as far as I'm aware), yet `===` can compare them perfectly fine, even though they're not mutable:

```
julia> primitive type Foo 8 end

julia> foo = reinterpret(Foo, 0x0)
Foo(0x00)

julia> fieldnames(Foo)
()

julia> ismutable(foo)
false

julia> bar = reinterpret(Foo, 0x1)
Foo(0x01)

julia> foo === bar
false

julia> foo === reinterpret(Foo, 0x0)
true
```

That is to say, the original explanation is correct - for non-mutable values, the bitwise contents matter, since those objects are _defined_ by their bitwise identity.

**Comment by gafter** [comment]
@Seelengrab I suspect you haven't read the latest revision.  It distinguishes immutable types from "builtin" types such as `Int`.  Do you have a specific proposal of what to change to address your concerns?

**Comment by oscardssmith** [comment]
the problem is that that distinction doesn't really exist as such. The only distinction that can be drawn is between `primitive type`s and other immutable types. (`Int` isn't any more builtin than any other primitive type).

**Comment by Seelengrab** [comment]
I have read the latest revision, which is what prompted my second response in the first place :) In particular, I think singleing out `String`s vs. other immutables here:

> and immutable objects (such as tuples and non-mutable
structs) are compared field-by-field using `===`. It is specialized for some types, such
as String, that should be treated as the same when the contents are the same.

doesn't really aid understanding. Semantically, at some point, there just aren't any fields left to compare/recurse on - an `Int` doesn't have fields, and neither does `String`. Since both are immutable, they are compared by using their bitwise contents, and not their location in memory (which is what would happen for mutables). How exactly a `String` is stored in another immutable is an implementation detail :shrug: 

My example with `primitive type` was just to show that you can have egality comparison without having any fields, which is what `===` on non-mutables is all about (`Int` or `String` are, in that regard, the same - `fieldnames` gives the empty tuple for both of them).

**Comment by Seelengrab** [comment]
I don't follow how `integer` makes a difference to my comments about `String` not being a special thing in regards to `===`..?

**Comment by gafter** [comment]
@Seelengrab: You wrote

> doesn't really aid understanding. Semantically, at some point, there just aren't any fields left to compare/recurse on - an Int doesn't have fields

The concept of a primitive type is part of the [Julia language documentation](https://docs.julialang.org/en/v1/manual/types/#Primitive-Types).  `String` isn't one of these.  Its classification in the documentation isn't clear, but it definitely gets special behavior and the proposed documentation calls that out.  If it were merely an immutable type with no fields, all strings would be equal by the specification (because all zero fields are `==`).  But strings are treated specially.

Perhaps the latest revision is better from your point of view?

**Comment by Seelengrab** [comment]
Semantically, a `String` is an immutable byte array whose entries are interpreted as a UTF-8 encoded sequence of `Char`s (see [here](https://github.com/JuliaLang/julia/blob/6bfb7fb13c3c60cbef17a87f49ea97b491afe792/base/strings/string.jl#L34-L44)). Internally, the length is also stored at the beginning of that array, making a string definitely a composite type and not a primitive type (not to mention that all instances of a primitive type have the same length, which isn't the case for `String`). This is also reflected in `isstructtype(String) === true`. That these internals aren't exposed is mostly for historical reasons - we don't have an immutable byte array in the frontend after all (there have been some attempts to get that though, like #48728 - having that would be very useful indeed).

I don't think singling out `String` here, especially without also having the underlying explanation of why this is so, is a good thing. It leads to more questions for a reader than answers, since all that matters for the question of "what does `===` do" is whether the objects are _semantically_ immutable and whether their _semantic_ contents are identical. This is already reflected in the current docstring, which is only concerned with the (semantic) contents of the objects, _not_ the implementation details of being stored as a pointer. Hence, for the purposes of `===`, primitive types and immutable struct types are the same. On this level, there is no meaningful distinction between primitive types and immutable struct types.

I think we need a better section on "immutable vs. mutable objects" in the manual, clarifying that both `primitive type`s and `struct` are immutable for all semantic purposes :thinking:

**Comment by gafter** [comment]
@Seelengrab Would you please propose a revision that addresses your concerns?

**Comment by Seelengrab** [comment]
As mentioned above, I'm not sure it necessarily should change at all. As I keep saying - the fact that under the hood, a pointer is stored in the tuple `("a", 1)` to refer to the string is an implementation detail. This is, I think, ultimately the misunderstanding that spurred the PR after all.

**Comment by ctarn** [comment]
The commit in orginal issue can be helpful: https://github.com/JuliaLang/julia/issues/49254#issuecomment-1497892175.

**Comment by gafter** [comment]
@ctarn that is one of the issues here.  String is an immutable structure and yet has no fields.

**Comment by mbauman** [comment]
Does having these details in the docs actually help the understanding of the function in the first place?  I think things are getting lost in the weeds here.

What about simply:

> Determine whether `x` and `y` are identical, in the sense that no program could distinguish
> them.
>
> Concretely, two things are `===` if and only if they describe the exact same object. That
> means that they must have the same type and value and they cannot change (mutate)
> independently of each other.

Not thrilled with "things" there, but I think it's better.

**Comment by ctarn** [comment]
I think we all (here) understand what `===` or `egal` is designed to be. I paste the original doc as below.

> Determine whether x and y are identical, in the sense that no program could distinguish them. First the types of x and y are compared. If those are identical, mutable objects are compared by address in memory and immutable objects (such as numbers) are compared by contents at the bit level. This function is sometimes called "egal". It always returns a Bool value.

The first sentence has clear defined what it is, or what it is designed to be. For this purpose, I think it's enough, although it also can be better. I will show my opinion on it at the end.

However, the single sentence may not help people understand the concept, e.g., what is `egal`? And thus the rest sentences were added to help understanding by discussing types, and immutable or mutable objects. As @gafter pointed out in https://github.com/JuliaLang/julia/issues/49254, comparing the content by bit level can also be misleading since some one may believe the pointer is part of the `content` instead of the object pointed to. The idea is reasonable. Actually, there is already some doc clarifying the problem:

> If all the fields of an immutable structure are indistinguishable (===) then two immutable values containing those fields are also indistinguishable: ...

Unfortunately above sentence only is located at doc about `Composite Types` (https://docs.julialang.org/en/v1/manual/types/#Composite-Types). People may not be able to notice this.

I believe that more descriptions can help people understand the concept, and agree that add above sentences and examples to the doc of `===` too. We don't have to introduce extra complexity or details but only helpful examples.

At the end, I suggest that changing the first sentence

> Determine whether x and y are identical, in the sense that no program could distinguish them.

into

> Determine whether x and y are semantically identical, in the sense that no program could distinguish them regardless of technical details like memory addresses.

, which should be more precise.

**Comment by Seelengrab** [comment]
Mentioning a memory address here that, semantically, doesn't exist is only going to lead to more questions, which is what I'm objecting to. A `Tuple{String, Int}` (from the original issue) does not semantically have a pointer (and which might not even exist at runtime) - otherwise it would be a `Tuple{Ptr{String}, Int}` or `Tuple{Ref{String}, Int}`. It's all about the contents of the tuple, which is a `String` and an `Int`, which need to be the same when doing a bitwise comparison. Semantically, due to being immutable, a `String` doesn't have a well-defined memory address (although in practice it does due to this allowing for some memory savings by reusing constants - this however is not relevant to `===`, since two string objects with the same bitwise content, though at different memory locations, are still considered to be the same string).

This is the core difference I'm talking about - semantically, the "contents" mentioned in the docstring are all about the types and what those types imply about the underlying bitwise representation & its interpretation, _not_ about how they're actually implemented. That's not the job of `===`, which is about semantic equality. For any immutable (`primitive type` or immutable `struct`) that simply refers to bitwise equality, nothing more and nothing less, since memory addresses _do not matter_ for those objects.

I recognize that this is kind of subtle in how the edge cases work in detail if you don't already have a good understanding of how the underlying systems works & behaves, but muddling the documentation by giving examples that are only superficially more correct than the true behavior, in my opinion, hurts more than it helps. As such, I really do think this means we need a better section about `Immutability vs. Mutability` in the manual, rather than adding a sentence or two to the docstring of `===`.

---

Perhaps adding something like

```
Note that both `primitive type`s and immutable `struct`s are semantically immutable, and both are compared by their bitwise contents.
```

could be helpful, but anything more than that (i.e. singling out specific examples), in my opinion, goes too far and is going to lead to confusion when the underlying implementation doesn't quite match the expected behavior.

**Comment by ctarn** [comment]
> A Tuple{String, Int} (from the original issue) does not semantically have a pointer (and which might not even exist at runtime) - otherwise it would be a Tuple{Ptr{String}, Int} or Tuple{Ref{String}, Int}.

Just as a supplement, something like a pointer does exist at runtime. A _pointer_ is stored in the tuple instead of copying the data every time. (Only for current implementation of course.)

Run:
```
a = "text"
println(pointer(a))
t1 = (a, 1)
println(pointer(t1[1]))
t2 = (a, 2)
println(pointer(t2[1]))
```

Output:
```
Ptr{UInt8} @0×00007f7fb0159ad8
Ptr{UInt8} @0×00007f7fb0159ad8
Ptr{UInt8} @0×00007f7fb0159ad8
```

**Review by ctarn — COMMENTED** [review]
Is the grammar of "A is ..., B is ..., and C is ..." correct? I am not a native speaker, so it is just a question :). I prefer "A is ...; B is ...; and C is ...".
Additionally, how about changing the example below to `a = [1, "string"]; b = [1, "string"];`?

**Comment by Seelengrab** [comment]
> Run:

None of these examples show that a pointer is stored in the tuple. They only show that, when you pass a `String` to `pointer`, you get a `Ptr` back. That the string came from a tuple is irrelevant to `pointer`, and in fact, `pointer` isn't even aware of that. Semantically, the tuple only has a `String` and an `Int`, no pointers involved.

**Pull request closed without merging**